pkg/reduce/data_forward_test.go

```diff
@@ -1095,6 +1108,103 @@ func TestDataForward_WithContextClose(t *testing.T) {
+		pipelineName   = "test-reduce-pipeline"
+		err            error
+	)
+
+	defer cancel()
+
+	// create from buffers
+	fromBuffer := simplebuffer.NewInMemoryBuffer(fromBufferName, fromBufferSize, 0)
+
+	// create to buffers
+	buffer1 := simplebuffer.NewInMemoryBuffer(toVertexName+"0", toBufferSize, 0)
+	buffer2 := simplebuffer.NewInMemoryBuffer(toVertexName+"1", toBufferSize, 1)
+	toBuffer := map[string][]isb.BufferWriter{
+		toVertexName: {buffer1, buffer2},
+	}
+
+	// create pbq manager
+	var pbqManager *pbq.Manager
+	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
+	assert.NoError(t, err)
+
+	// create in memory watermark publisher and fetcher
+	f, p := fetcherAndPublisher(ctx, fromBuffer, t.Name())
+	publishersMap, _ := buildPublisherMapAndOTStore(ctx, toBuffer, pipelineName)
+	// create a fixed window of 5 minutes
+	window := fixed.NewFixed(5 * time.Minute)
+
+	idleManager := wmb.NewIdleManager(len(toBuffer))
+	op := pnf.NewOrderedProcessor(ctx, keyedVertex, SumReduceTest{}, toBuffer, pbqManager, &myForwardTestRoundRobin{}, publishersMap, idleManager)
+
+	var reduceDataForward *DataForward
+	reduceDataForward, err = NewDataForward(ctx, keyedVertex, fromBuffer, toBuffer, pbqManager, &myForwardTestRoundRobin{}, f, publishersMap,
+		window, idleManager, op, WithReadBatchSize(10))
+
+	assert.NoError(t, err)
+
+	// start the producer
+	go publishMessages(ctx, startTime, messages, 650, 10, p, fromBuffer)
+
+	// start the forwarder
+	go reduceDataForward.Start()
+
+	msgs0, readErr := buffer1.Read(ctx, 1)
+	assert.Nil(t, readErr)
+	for len(msgs0) == 0 || msgs0[0].Header.Kind == isb.WMB {
+		select {
+		case <-ctx.Done():
+			assert.Fail(t, ctx.Err().Error())
+			return
+		default:
+			time.Sleep(100 * time.Millisecond)
+			msgs0, readErr = buffer1.Read(ctx, 1)
+			assert.Nil(t, readErr)
+		}
+	}
+
+	msgs1, readErr1 := buffer2.Read(ctx, 1)
+	assert.Nil(t, readErr1)
+	for len(msgs1) == 0 || msgs1[0].Header.Kind == isb.WMB {
+		select {
+		case <-ctx.Done():
+			assert.Fail(t, ctx.Err().Error())
+			return
+		default:
+			time.Sleep(100 * time.Millisecond)
+			msgs1, readErr1 = buffer2.Read(ctx, 1)
+			assert.Nil(t, readErr1)
+		}
+	}
+
+	// assert the output of reduce
+	var readMessagePayload0 PayloadForTest
+	_ = json.Unmarshal(msgs0[0].Payload, &readMessagePayload0)
+	// since the window duration is 5 minutes, the output should be
+	// 100 * 300(for key even) and 99 * 300(for key odd)
+	// we cant guarantee the order of the output
+	assert.Contains(t, []int{30000, 29700}, readMessagePayload0.Value)
+	assert.Contains(t, []string{"even", "odd"}, readMessagePayload0.Key)
+
+	var readMessagePayload1 PayloadForTest
+	_ = json.Unmarshal(msgs1[0].Payload, &readMessagePayload1)
+	assert.Contains(t, []int{30000, 29700}, readMessagePayload1.Value)
+	assert.Contains(t, []string{"even", "odd"}, readMessagePayload1.Key)
```

Comment:
We initialized both `ordered processor` and `reduce data forwarder` with `myForwardTestRoundRobin`. Should we expect the reduce vertex output to be send to toBuffers round-robinly? Why do we expect that both buffer contain same data? Is it because they reflect different windows but same window size?

Reply:
yes because we are producing more than one message, so we are expecting both buffers to have some data.